uniffi_bindgen/src/bindings/kotlin/templates/ObjectRuntime.kt

```diff
@@ -71,13 +74,23 @@
 // called *and* all in-flight method calls have completed, avoiding violating any of the expectations
 // of the underlying Rust code.
 //
-// In the future we may be able to replace some of this with automatic finalization logic, such as using
-// the new "Cleaner" functionaility in Java 9. The above scheme has been designed to work even if `destroy` is
-// invoked by garbage-collection machinery rather than by calling code (which by the way, it's apparently also
-// possible for the JVM to finalize an object while there is an in-flight call to one of its methods [1],
-// so there would still be some complexity here).
+// The `destroy` method is implemented in terms of a `Cleaner.Cleanable`, and it is a `CleanAction`
+// that actually calls into to Rust. In this manner, we're able repeat whatever guarantees `java.ref.Cleaner`
+// makes about cleaning up these "native peers".
+//
+// A cleaner makes a better alternative to _not_ calling `destroy()` as
+// and when the object is finished with, but the abstraction is not perfect: if the Rust object's `drop`
+// method is slow, and/or there are many objects to cleanup, and it's on a low end Android device, then the cleaner
+// thread may be starved, and the app will leak memory.
+//
+// In this case, `destroy`ing manually may be a better solution.
 //
-// Sigh...all of this for want of a robust finalization mechanism.
+// The cleaner can live side by side with the manual calling of `destroy`. In the order of preference, uniffi objects
+// with Rust peers can be reclaimed:
+//
+// 1. By calling the `destroy` method of the object, which calls `rustObject.free()`. If that doesn't happen:
+// 2. When the object becomes unreachable, AND the Cleaner thread gets to call `rustObject.free()`. If the thread is starved then:
+// 3. The memory is reclaimed when the process terminates.
```

Comment:
I prefer relying on `2` and would love to get rid of the `destroy()` method.  It gets around [corner cases](https://github.com/mozilla/uniffi-rs/pull/1787#issuecomment-1788966499) and feels more consistent with GC-based language.

How serious is the risk of cleaner thread starvation?  I hope it's something most applications can ignore.

I don't think this needs to be a blocker.  I'd be happy to merge this one as-is and discuss in a future PR.

Reply:
(1) is the escape hatch for when thread starvation happens. As you say, most apps won't have to worry about it.

We should deprecate and then remove the `Disposable` interface, and the `destroy` method, though this isn't done here, because we should go through the formal deprecation process.